Mage.Sets/src/mage/cards/k/KorvoldGleefulGlutton.java

```diff
@@ -0,0 +1,168 @@
+        
+        this.addAbility(staticAbility, new KorvoldGleefulGluttonWatcher());
+
+        // Flying, Trample, Haste
+        this.addAbility(FlyingAbility.getInstance());
+        this.addAbility(TrampleAbility.getInstance());
+        this.addAbility(HasteAbility.getInstance());
+
+        // Whenever Korvold deals combat damage to a player, put X +1/+1 counters on Korvold and draw X cards, where X is the number of permanent types among cards in your graveyard.
+        Ability combatDamageAbility = new DealsCombatDamageToAPlayerTriggeredAbility(
+            new AddCountersSourceEffect(CounterType.P1P1.createInstance(), PermanentTypesInGraveyardCount.instance, true),
+        false
+        );
+        combatDamageAbility.addEffect(new DrawCardSourceControllerEffect(PermanentTypesInGraveyardCount.instance).setText("and draw X cards, where X is the number of permanent types among cards in your graveyard"));
+        combatDamageAbility.addHint(new ValueHint("permanent types among cards in your graveyard", PermanentTypesInGraveyardCount.instance));
+        this.addAbility(combatDamageAbility);
+    }
+
+    private KorvoldGleefulGlutton(final KorvoldGleefulGlutton card) {
+        super(card);
+    }
+
+    @Override
+    public KorvoldGleefulGlutton copy() {
+        return new KorvoldGleefulGlutton(this);
+    }
+}
+
+class KorvoldGleefulGluttonWatcher extends Watcher {
+
+    private final Map<UUID, Set<CardType>> map = new HashMap<>();
+
+    KorvoldGleefulGluttonWatcher() {
+        super(WatcherScope.GAME);
+    }
+
+    @Override
+    public void watch(GameEvent event, Game game) {
+        if (event.getType() != GameEvent.EventType.SACRIFICED_PERMANENT) {
+            return;
+        }
+
+        Permanent permanent = game.getPermanentOrLKIBattlefield(event.getTargetId());
+
+        if (permanent != null) {
+            permanent.getCardType(game).forEach(type -> 
+                map.computeIfAbsent(event.getPlayerId(), (key) -> new HashSet<>()).add(type)
+            );
+        }
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        map.clear();
+    }
+
+    public int getNumberOfTypes(UUID playerId) {
+        return map.computeIfAbsent(playerId, (key) -> new HashSet<>()).size();
+    };
+}
+
+enum CardTypesAmongSacrifiedPermanents implements DynamicValue {
+    instance;
+
+    @Override
+    public int calculate(Game game, Ability sourceAbility, Effect effect) {
+        KorvoldGleefulGluttonWatcher watcher = game.getState().getWatcher(null);
+        return watcher == null ? 0 : watcher.getNumberOfTypes(sourceAbility.getControllerId());
+    }
+
+    @Override
+    public DynamicValue copy() {
+        return instance;
+    }
+
+    @Override
+    public String getMessage() {
+        return "card types among permanents you sacrificed";
+    }
+}
+
+enum PermanentTypesInGraveyardCount implements DynamicValue {
```

Comment:
I prefer naming card-specific classes with the card name so it's obvious they aren't common classes

Reply:
I wanted to keep them clearly named, and I figured adding `KorvoldGleefulGlutton` to the beginning would make the names too wordy